ltdconveyor/s3upload.py

```diff
@@ -0,0 +1,395 @@
+    bucket, and can delete objects that no longer exist in the source.
+
+    Parameters
+    ----------
+    session : :class:`boto3.session.Session`
+        A boto3 session instance provisioned with the correct identities.
+    bucket_name : `str`
+        Name of the S3 bucket.
+    bucket_root : `str`
+        The version slug is the name root directory in the bucket where
+        documentation is stored.
+    """
+    def __init__(self, session, bucket_name, bucket_root):
+        super().__init__()
+        s3 = session.resource('s3')
+        bucket = s3.Bucket(bucket_name)
+        self._session = session
+        self._bucket = bucket
+        self._bucket_root = bucket_root
+        # Strip trailing '/' from bucket_root for comparisons
+        if self._bucket_root.endswith('/'):
+            self._bucket_root = self._bucket_root.rstrip('/')
+
+    def list_filenames_in_directory(self, dirname):
+        """List all file-type object names that exist at the root of this
+        bucket directory.
+
+        Parameters
+        ----------
+        dirname : `str`
+            Directory name in the bucket relative to ``bucket_root/``.
+
+        Returns
+        -------
+        filenames : `list`
+            List of file names (`str`), relative to ``bucket_root/``, that
+            exist at the root of ``dirname``.
+        """
+        prefix = self._create_prefix(dirname)
+        filenames = []
+        for obj in self._bucket.objects.filter(Prefix=prefix):
+            if obj.key.endswith('/'):
+                continue
+            obj_dirname = os.path.dirname(obj.key)
+            if obj_dirname == prefix:
+                # object is at root of directory
+                filenames.append(os.path.relpath(obj.key,
+                                                 start=prefix))
+        return filenames
+
+    def list_dirnames_in_directory(self, dirname):
+        """List all names of directories that exist at the root of this
+        bucket directory.
+
+        Note that *directories* don't exist in S3; rather directories are
+        inferred from path names.
+
+        Parameters
+        ----------
+        dirname : `str`
+            Directory name in the bucket relative to ``bucket_root``.
+
+        Returns
+        -------
+        dirnames : `list`
+            List of directory names (`str`), relative to ``bucket_root/``,
+            that exist at the root of ``dirname``.
+        """
+        prefix = self._create_prefix(dirname)
+        dirnames = []
+        for obj in self._bucket.objects.filter(Prefix=prefix):
+            dirname = os.path.dirname(obj.key)
+            # if the object is a directory redirect, make it look like a dir
+            if dirname == '':
+                dirname = obj.key + '/'
+            rel_dirname = os.path.relpath(dirname, start=prefix)
+            dir_parts = rel_dirname.split('/')
+            if len(dir_parts) == 1:
```

Comment:
I suggest adding a comment here explaining the conditional.  I had to really had to think about this (because if the path has multiple parts, it must not be at this "level" of the bucket).